tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,387 @@
+from typing import List, Optional, Union, Tuple
+import itertools
+from typing_extensions import Literal
+from dataclasses import dataclass
+import textwrap
+from tools.codegen import local
+from tools.codegen.context import method_with_native_function, native_function_manager
+from tools.codegen.utils import Target, mapMaybe
+from tools.codegen.model import (BaseType, OptionalType, DispatchKey, NativeFunction,
+                                 NativeFunctionsGroup, SchemaKind, FunctionSchema,
+                                 TensorOptionsArguments, ListType,
+                                 DeviceCheckType, Argument, assert_never,
+                                 is_cuda_dispatch_key, BackendIndex,
+                                 gets_generated_out_inplace_wrapper, OperatorName,
+                                 Arguments, SelfArgument, Return)
+from tools.codegen.api.types import (BaseTy, BaseCppType, BaseCType, OptionalCType,
+                                     Binding, ConstRefCType, NamedCType,
+                                     CppSignature, CppSignatureGroup,
+                                     Expr, MutRefCType, kernel_signature,
+                                     DispatcherSignature, VectorCType, intT, ListCType,
+                                     scalarT, scalarTypeT)
+import tools.codegen.api.meta as meta
+import tools.codegen.api.cpp as cpp
+import tools.codegen.api.structured as structured
+from tools.codegen.api.translate import translate
+from tools.codegen.selective_build.selector import SelectiveBuilder
+
+valueListT = BaseCppType('std', 'vector<ir::Value>')
+valueT = BaseCppType('ir', 'Value')
+intArrayT = BaseCppType('std', 'vector<int64_t>')
+
+def process_ir_type(typ: BaseTy, mutable: bool, binds: str) -> NamedCType:
+    """
+    This function takes a type from NativeFunctions and converts it for use with
+    lazy tensor codegen.  Currently its output is used in several places, and so far
+    it has been possible for them to all use the same conversions, but that may not be
+    optimal or possible in the finished system.
+
+    Type conversion for lazy currently consists of
+     (1) changing Tensor-like things into Value-like things
+     (2) wrapping everythign in a BaseCType
+     (3) wrapping everything further in a NamedCType
+
+    (1) converts Tensors to Values since Values are how Lazy IR represents tensors.  There
+    is special handling for Optional[Tensor] or List[Tensor], etc- hence 'tensor-like'
+
+    This is incomplete- there are assertions in places that it's expected to need to add
+    more types as the codegen is used with more operators.
+    """
+    if isinstance(typ, BaseType):
+        if typ.name == BaseTy.Tensor:
+            return NamedCType(binds, BaseCType(valueT))
+        elif typ.name == BaseTy.Scalar:
```

Comment:
This function is a no-op if `not type.is_tensor_like()`, right? Maybe that would simplify some of this logic (you don't need to deal with all of the non-tensor cases)

Reply:
@ezyang actually steered me away from calling back into argumenttype_type and encouraged me to reimplement the whole logic here.  I ended up agreeing that this way is a bit clearer.